docs/http-cloud-provider.md

```diff
@@ -0,0 +1,643 @@
+
+# Specification
+
+The user should specify a config file in a INI config-file format. The file is read using gcfg. The **client-url** field is required which specifies the URL of the cloud provider. The client-url field should come under the Global section.
+
+##### Example Config File:
+
+```bash
+[Global]
+client-url = http://127.0.0.1 
+```
+
+First the API checks for which interfaces in [cloud.go](../pkg/cloudprovider/cloud.go) are implemented.
+
+A GET request for /v1/{Interface}/supported will return whether the interface is supported or not.
+
+The following methods must be implemented by the HTTP cloudprovider. If a cloudprovider does not implement a resource, it can return false  or return a HTTP 404 Error with empty body for a given GET /{interface}/supported call.
+
+If you encounter an error in processing the request, return an HTTP 404 with a response body of:
+
+```json
+{
+    "error" :"error-message-here"
+}
+```
+
+## Provider Name
+
+#### GET /v1/providerName
+
+Returns the cloud provider name.
+
+##### Expected return value:
+
+|Name|Description|Example Value|
+|----|-----------|-------------|
+|providerName|Name of the cloud provider.|"my-cloud"|
+Example
+
+```json
+{
+    "providerName": "my-cloud"
+}
+```
+
+## Instances
+
+The instances interface allows the kubernetes platform to query the cloud provider about the instances currently within the cloud provider system.
+
+### Methods
+
+#### GET /v1/instances/supported
+
+Returns whether the Instance interface is supported or not.
+
+##### Expected return value:
+
+|Name|Description|Example Value|
+|----|-----------|-------------|
+|supported|Whether the interface is supported or not.|true|
+Example
+
+```json
+{
+    "supported": true
+}
+```
+
+#### GET /v1/instances/{instanceName}/nodeAddresses
+
+Returns all node addresses for the given instance.
+
+##### Expected return value:
+
+|Name|Description|Example Value|
+|----|-----------|-------------|
+|nodeAddresses|List of node adress objects.||
+|nodeAddresses:type|The IP type relate to the node address. The types of addresses are restricted to the list below.|"InternalIP"|
+|nodeAddresses:address|The actual node's address.|"127.0.0.1"|
+
+Address Types
+
+|Name|Description|
+|----|-----------|
+|InternalIP|The node's Internal IP address.|
+|ExternalIP|The node's External IP address.|
+|Hostname|The hostname of the node.|
+|LegacyHostIP|The node's Host IP address.|
+Example
+
+```json
+{
+    "nodeAddresses": [
+        {
+            "type": "InternalIP",
+            "address": "127.0.0.1"
+        }
+    ]
+}
+```
+
+#### GET /v1/instances/{instanceName}/ID
```

Comment:
it seems like GET /v1/instances/{instanceName}/nodeAddresses and GET /v1/instances/{instanceName}/ID should be merged into a single GET /v1/instances/{instanceName}. I don't see much value in per-field resources like these. Having a one-to-one mapping of cloudprovider interface method to resource doesn't seem worth it.